google/cloud/internal/retry_policy.cc

```diff
@@ -0,0 +1,41 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/retry_policy.h"
+#include "absl/strings/match.h"
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+
+bool IsTransientInternalError(Status const& status) {
+  // Treat the unexpected termination of the gRPC connection as retryable.
+  // There is no explicit indication of this, but it will result in an
+  // INTERNAL status with one of the `kTransientFailureMessages`.
+  if (status.code() == StatusCode::kInternal) {
+    static constexpr char const* kTransientFailureMessages[] = {
+        "RST_STREAM", "Received Rst Stream",
+        "Received unexpected EOS on DATA frame from server"};
+    for (auto const& message : kTransientFailureMessages) {
```

Comment:
nit: consider

```cc
  if (status.code() != Statuscode::kInternal) return false;
  return std::any_of(
      std::begin(kTransienFailureMessages),
      std::end(kTransientFailureMessages),
     [&status](char const* message) {
        return absl::StrContains(status.message(), message);
   });
```

Reply:
Updated. I did consider an uglier version of it.